cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,292 @@
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const T1& a, const T2& b) const
+  {
+    return a / b;
+  }
+};
+
+struct div_checkzero_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (b == Type{0}) { return Type{0}; }
+    return a / b;
+  }
+};
+
+struct pow_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    return std::pow(a, b);
+  }
+};
+
+struct min_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (a > b) { return b; }
+    return a;
+  }
+};
+
+struct max_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+
+struct sqdiff_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    Type diff = a - b;
+    return diff * diff;
+  }
+};
+
+struct argmin_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr const_op(const ScalarT& s) : scalar{s} {}
+
+  template <typename... Args>
+  constexpr RAFT_INLINE_FUNCTION ScalarT operator()(Args...) const
+  {
+    return scalar;
+  }
+};
+
+template <typename ScalarT, typename BinaryOpT>
+struct scalar_op {
+  const ScalarT scalar;
+  const BinaryOpT composed_op;
+
+  template <typename OpT     = BinaryOpT,
+            typename UnusedT = std::enable_if_t<std::is_default_constructible_v<OpT>>>
+  constexpr scalar_op(const ScalarT& s) : scalar{s}, composed_op{}
+  {
+  }
+  constexpr scalar_op(const ScalarT& s, BinaryOpT o) : scalar{s}, composed_op{o} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(InT a) const
+  {
+    return composed_op(a, scalar);
+  }
+};
+
+template <typename Type>
+using scalar_add_op = scalar_op<Type, add_op>;
+
+template <typename Type>
+using scalar_sub_op = scalar_op<Type, sub_op>;
+
+template <typename Type>
+using scalar_mul_op = scalar_op<Type, mul_op>;
+
+template <typename Type>
+using scalar_div_op = scalar_op<Type, div_op>;
+
+template <typename Type>
+using scalar_div_checkzero_op = scalar_op<Type, div_checkzero_op>;
+
+template <typename Type>
+using scalar_pow_op = scalar_op<Type, pow_op>;
+
+template <typename OuterOpT, typename InnerOpT>
```

Comment:
I saw we'd got at least one case where we nest several `compose_op`s. Then, bonus points for making this arbitrary chain composition :)
```C++
template <typename... OpsT>
struct compose_op {
  const std::tuple<OpsT...> ops;

  constexpr explicit compose_op(OpsT... ops) : ops(ops...) {}

  template <typename... Args>
  constexpr RAFT_INLINE_FUNCTION auto operator()(Args&&... args) const
  {
    return compose<sizeof...(OpsT)>(std::forward<Args>(args)...);
  }

 private:
  template <size_t RemOps, typename... Args>
  constexpr RAFT_INLINE_FUNCTION auto compose(Args&&... args) const
  {
    if constexpr (RemOps > 0) {
      return compose<RemOps - 1>(std::get<RemOps - 1>(ops)(std::forward<Args...>(args)...));
    } else {
      return identity_op{}(std::forward<Args...>(args)...);
    }
  }
};
```
Then, the composition in `ivf_pq_search:434` would look much simpler:
```C++
compose_op{scalar_mul_op<float>{scaling_factor}, sqrt_op{}, cast_op<float>{}}
```